src/fastaq_handler.cpp

```diff
@@ -1,176 +1,83 @@
 #include <string>
-#include <cstring>
 #include <iostream>
-//#include <fstream>
-#include <boost/iostreams/filtering_streambuf.hpp>
-#include <boost/iostreams/copy.hpp>
-#include <boost/iostreams/filter/gzip.hpp>
 #include "fastaq_handler.h"
-#include "utils.h"
 
-FastaqHandler::FastaqHandler(const std::string& filepath)
-    : gzipped(false)
-    , instream(&inbuf)
+FastaqHandler::FastaqHandler(const std::string filepath)
+    : closed(false)
+    , filepath(filepath)
     , num_reads_parsed(0)
 {
-    // level for boost logging
-    //    logging::core::get()->set_filter(logging::trivial::severity >= g_log_level);
-
-    BOOST_LOG_TRIVIAL(debug) << "Open fastaq file" << filepath;
-    fastaq_file.open(filepath);
-    if (not fastaq_file.is_open()) {
-        std::cerr << "Unable to open fastaq file " << filepath << std::endl;
-        std::exit(EXIT_FAILURE);
-    }
-    try {
-        if (filepath.substr(filepath.length() - 2) == "gz") {
-            inbuf.push(boost::iostreams::gzip_decompressor());
-            gzipped = true;
-        }
-        inbuf.push(fastaq_file);
-    } catch (const boost::iostreams::gzip_error& e) {
-        std::cerr << "Problem transfering file contents to boost stream: " << e.what()
-                  << '\n';
+    this->fastaq_file = gzopen(filepath.c_str(), "r");
+    if (this->fastaq_file == nullptr) {
+        throw "Unable to open " + this->filepath;
```

Comment:
C++ allows us to throw whatever object as an exception, but it is nice to throw subclasses of `std::exception` (https://en.cppreference.com/w/cpp/error/exception). `std::runtime_error` is a very general one, and can fit all our cases. The issue on throwing objects that are not subclasses of `std::exception` is that if we error out on an exception (e.g. the given fasta/q file does not exist), C++ will error out with this message (in this example, as we are throwing `std::string`):
```
terminate called after throwing an instance of 'std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >'
```
`std::string` is an alias for `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >`. Anyway, all that C++ says is that an exception of type `std::string` or whatever other type was thrown, but it does not describe the error itself. If we instead `throw std::runtime_error("Unable to open " + filepath);`, we get this when a file does not exist:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Unable to open reads.fa
```

The first line is identical as before, just outputting the type of the exception, but then C++ sees that the object thrown is an `std::exception`, and calls the `what()` method, which prints the string given to the object at its construction. Then the user can understand the error.

You could also build your own exception classes for each type of exception your code might throw, but I don't think this is necessarily required here... just throwing an object that is a `std::exception`, either from std library or your own class is enough

Reply:
Yep, totally agree. This was me being lazy.